sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -203,3 +203,23 @@ abstract class BinaryNode extends LogicalPlan {
 abstract class OrderPreservingUnaryNode extends UnaryNode {
   override final def outputOrdering: Seq[SortOrder] = child.outputOrdering
 }
+
+trait LogicalPlanIntegrity {
```

Comment:
If just for `hasUniqueIdsForAttributes`, a `trait` seems too much. Maybe just as an object `LogicalPlanIntegrity`?